File: Design-Patterns-Using-CSharp/ChainOfResponsibility/ChainOfResponsibilityGUI.cs
using System;

// Abstract class representing a handler
abstract class Handler
{
    protected Handler parent;
    protected string helpText;

    public Handler(Handler parent)
    {
        this.parent = parent;
        this.helpText = null;
    }

    public abstract void ShowHelperText();

    public void SetHelperText(string text)
    {
        this.helpText = text;
    }
}

// Concrete class representing a container handler
class Container : Handler
{
    public Container(Handler parent) : base(parent)
    {
    }

    public override void ShowHelperText()
    {
        if (helpText != null)
        {
            Console.WriteLine("Help :: " + helpText);
        }
        else if (parent != null)
        {
            Console.WriteLine("Message passed to next in chain by Container");
            parent.ShowHelperText();
        }
    }
}

// Concrete class representing a button handler
class Button : Handler
{
    private string label;

    public Button(string label, Handler parent) : base(parent)
    {
        this.label = label;
    }

    public override void ShowHelperText()
    {
        if (helpText != null)
        {
            Console.WriteLine("Help :: " + helpText);
        }
        else if (parent != null)
        {
            Console.WriteLine("Message passed to next in chain by Button");
            parent.ShowHelperText();
        }
    }
}

// Concrete class representing a panel handler
class Panel : Handler
{
    public Panel() : base(null)
    {
    }

    public override void ShowHelperText()
    {
        if (helpText != null)
        {
            Console.WriteLine("Help :: " + helpText);
        }
        else if (parent != null)
        {
            Console.WriteLine("Message passed to next in chain by Panel");
            parent.ShowHelperText();
        }
    }
}

// Client code
public class ChainOfResponsibilityGUI
{
    public static void Main(string[] args)
    {
        Panel p = new Panel();
        p.SetHelperText("Panel help text.");

        Button b1 = new Button("Ok", p);
        b1.SetHelperText("Ok button help text.");

        Button b2 = new Button("Cancel", p);

        b1.ShowHelperText();
        b2.ShowHelperText();
    }
}

File: Design-Patterns-Using-Swift/ChainOfResponsibility/ChainOfResponsibilityGUI.swift
// Abstract class representing a handler
class Handler {
    var parent: Handler?
    var helpText: String?

    init(parent: Handler?) {
        self.parent = parent
        self.helpText = nil
    }

    func showHelperText() {
        fatalError("Abstract method. Subclasses must implement this method.")
    }

    func setHelperText(text: String) {
        self.helpText = text
    }
}

// Concrete class representing a container
class Container: Handler {
    override func showHelperText() {
        if let helpText = helpText {
            print("Help :: \(helpText)")
        } else if let parent = parent {
            print("Message passed to next in chain by Container")
            parent.showHelperText()
        }
    }
}

// Concrete class representing a button
class Button: Handler {
    var label: String

    init(label: String, parent: Handler?) {
        self.label = label
        super.init(parent: parent)
    }

    override func showHelperText() {
        if let helpText = helpText {
            print("Help :: \(helpText)")
        } else if let parent = parent {
            print("Message passed to next in chain by Button")
            parent.showHelperText()
        }
    }
}

// Concrete class representing a panel
class Panel: Handler {
    override init(parent: Handler?) {
        super.init(parent: parent)
    }

    override func showHelperText() {
        if let helpText = helpText {
            print("Help :: \(helpText)")
        } else if let parent = parent {
            print("Message passed to next in chain by Panel")
            parent.showHelperText()
        }
    }
}

// Client code
let p = Panel(parent: nil)
p.setHelperText(text: "Panel help text.")

let b1 = Button(label: "Ok", parent: p)
b1.setHelperText(text: "Ok button help text.")

let b2 = Button(label: "Cancel", parent: p)

b1.showHelperText()
b2.showHelperText()

/*
Help :: Ok button help text.
Message passed to next in chain by Button
Help :: Panel help text.
*/